### PR TITLE
modified training.py to take multiple datasets + names the model in wandb to include the names of all datasets used

### PR DESCRIPTION
**Pull request by tonnyryam**

### CODE REVIEW

File: learning/training.py [hunk 1]
```diff
@@ -193,7 +199,7 @@ def x2_from_filename(filename) -> float:
     )
 
     return image_command_data.dataloaders(
```
Comment (anthonyjclark) [inline learning/training.py:201]:
```suggestion
    # TODO: This is untested
    return image_command_data.dataloaders(
```

File: learning/training.py [hunk 2]
```diff
@@ -237,9 +243,12 @@ def train_model(dls: DataLoaders, args: Namespace, run, rep: int):
 
     wandb_name = args.wandb_name
     model_arch = args.model_arch
-    dataset_name = args.dataset_name
+    if isinstance(args.dataset_names, str):
```
Comment (anthonyjclark) [inline learning/training.py:246]: Isn't `args.dataset_names` always a list?

**Review by anthonyjclark — CHANGES_REQUESTED** [review]

File: learning/inference.py [hunk 3]
```diff
@@ -12,6 +12,8 @@
 from fastai.callback.wandb import WandbCallback
 from fastai.vision.learner import load_learner
 from utils import y_from_filename  # noqa: F401 (needed for fastai load_learner)
+import ffmpeg
+import subprocess
```
Comment (anthonyjclark) [inline learning/inference.py:16]: What are these here for?

File: learning/inference.py [hunk 4]
```diff
@@ -98,9 +100,6 @@ def parse_args():
         default=10,
         help="Maximum number of actions to take.",
     )
-
-    add_box_navigator_arguments(arg_parser)
-
```
Comment (anthonyjclark) [inline learning/inference.py:103]: I think we need this.

File: learning/training.py [hunk 5]
```diff
@@ -237,9 +244,9 @@ def train_model(dls: DataLoaders, args: Namespace, run, rep: int):
 
     wandb_name = args.wandb_name
     model_arch = args.model_arch
-    dataset_name = args.dataset_name
+    dataset_names = "-".join(args.dataset_names)
 
-    learn_name = f"{wandb_name}-{model_arch}-{dataset_name}-rep{rep:02}"
+    learn_name = f"{wandb_name}-{model_arch}-{dataset_names}-rep{rep:02}"
```
Comment (anthonyjclark) [inline learning/training.py:249]: Do you think we need the dataset names as part of the file name? Maybe `wandb_name` is enough?

**Review by anthonyjclark — CHANGES_REQUESTED** [review]

File: learning/inference.py (hunk 3 above)
Comment (tonnyryam) [reply learning/inference.py:16]: I was trying to automate the ffmpeg video/gif making process with the photos inference produces. I believe subprocess can enter script into the command prompt.

File: learning/inference.py (hunk 4 above)
Comment (tonnyryam) [reply learning/inference.py:103]: Yes, I don't know why it wasn't there.

File: learning/inference.py (hunk 3 above)
Comment (anthonyjclark) [reply learning/inference.py:16]: Yes, we could do that. But I didn't see that code anywhere. Did I miss it?

File: learning/training.py (hunk 5 above)
Comment (tonnyryam) [reply learning/training.py:249]: It could be quicker to include it for documenting purposes if we produce more datasets later on. It would be quicker to have them automatically included without having to write them in the wandb_name argument. I went with it since it was always there but it's up to you.

File: learning/inference.py (hunk 3 above)
Comment (tonnyryam) [reply learning/inference.py:16]: No I have yet to implement it.

File: learning/inference.py [hunk 6]
```diff
@@ -179,7 +179,9 @@ def main():
     )
 
     pbar_manager = enlighten.get_manager()
-    trials_pbar = pbar_manager.counter(total=args.num_trials, desc="Trials: ")
+    trials_pbar = pbar_manager.counter(
```
Comment (anthonyjclark) [inline learning/inference.py:182]: I think the actions and navigation pbars can be sub-bars of trials_pbar. What do you think?

File: learning/inference.py [hunk 7]
```diff
@@ -269,6 +275,31 @@ def main():
     inference_data_table = wandb.Table(columns=table_cols, data=inference_data)
     run.log({"Inference Data": inference_data_table})
 
+    video_name = (str(agent.image_directory).split("/"))[-1]
```
Comment (anthonyjclark) [inline learning/inference.py:278]: Use pathlib instead of string manipulation. Check out [PurePath.stem](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.stem)

File: learning/inference.py [hunk 8]
```diff
@@ -269,6 +275,31 @@ def main():
...
+    filelist_name = video_name + "_filelist.txt"
+
+    image_files = sorted([f for f in agent.image_directory.iterdir()])
+    with open(filelist_name, "w") as file_out:
+        for file in image_files:
+            file_out.write(f"file '{Path(file)}'\n")
```
Comment (anthonyjclark) [inline learning/inference.py:284]: Use pathlib for this. Check out [Path.glob](https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob)

File: learning/inference.py [hunk 9]
```diff
@@ -269,6 +275,31 @@ def main():
...
+            "concat",
+            "-safe",
+            "0",
+            "-i",
+            filelist_name,
+            "-c:v",
+            "libx264",
+            "-pix_fmt",
+            "yuv420p",
+            agent.image_directory / (video_name + ".mp4"),
```
Comment (anthonyjclark) [inline learning/inference.py:299]: Are all of these necessary? Provide some documentation.

**Review by anthonyjclark — CHANGES_REQUESTED** [review]

File: learning/inference.py (hunk 6 above)
Comment (tonnyryam) [reply learning/inference.py:182]: Adding them as sub-bars could be helpful when visualizing the broader results of all of the trials rather than each individual trial. Like we could see that on average the agent completed 10% of the environment on average when navigation fills out 10% of trials progress bar. It could be confusing with actions as a sub-bar since the max actions can vary each time as we haven't decided on a standard number yet. I think all we want is as few actions as possible with highest completion. We can create a metric like completion % divided by # of actions, but I don't think actions really needs a progress bar.

**Comment by anthonyjclark** [comment]
Let me know when this is ready for review.

File: boxnav/boxsim.py [hunk 10]
```diff
@@ -1,9 +1,11 @@
 from argparse import ArgumentParser, Namespace
 from math import radians
-from pathlib import Path
+from pathlib import Path, PurePath
```
Comment (anthonyjclark) [inline boxnav/boxsim.py:3]: Why `PurePath`?

File: boxnav/boxsim.py [hunk 11]
```diff
@@ -30,7 +32,6 @@ def simulate(args: Namespace) -> None:
 
     pbar_manager = enlighten.get_manager()
     actions_pbar = pbar_manager.counter(total=args.max_total_actions, desc="   Actions")
-    navigation_pbar = pbar_manager.counter(total=100, desc="Completion")
```
Comment (anthonyjclark) [inline boxnav/boxsim.py:33]: Why did you remove the navigation percentage bar?

File: boxnav/boxsim.py [hunk 12]
```diff
@@ -56,20 +57,36 @@ def simulate(args: Namespace) -> None:
...
+        sprun(
+            [
+                "python",
+                "upload_data.py",
+                PurePath(agent.image_directory).stem,
+                args.auto_upload,
+                (
+                    "Automatically uploading data from "
+                    + PurePath(agent.image_directory).stem
+                    + " run."
```
Comment (anthonyjclark) [inline boxnav/boxsim.py:82]: Use an f-string.

File: boxnav/boxsim.py [hunk 13]
```diff
@@ -56,20 +57,36 @@ def simulate(args: Namespace) -> None:
...
         agent.ue.close_osc()
 
-    navigation_pbar.close()
     actions_pbar.close()
 
     print("Simulation complete.")
 
+    if args.auto_upload:
+        chdir("..")
+        chdir("learning")
```
Comment (anthonyjclark) [inline boxnav/boxsim.py:72]: Can this be done with one call to `chdir`?

File: learning/inference.py [hunk 14]
```diff
@@ -7,16 +7,19 @@
 from contextlib import contextmanager
 from functools import partial
 from math import radians
-from pathlib import Path
+from pathlib import Path, PurePath
```
Comment (anthonyjclark) [inline learning/inference.py:10]: Why `PurePath`?

File: learning/inference.py [hunk 15]
```diff
@@ -453,6 +484,41 @@ def main():
...
+            "0",
+            # "-i" is the input file argument and "filelist_name" is the file containing the paths to the images that will be concatenated
+            "-i",
+            filelist_name,
+            # "-c:v" sets the codec and specifies it is for video stream and "libx264" specifies the codec to encode the video stream
+            # "-c:v",
+            # "libx264",
+            # # "-pix_fmt" specifies the pixel format for the output video and "yuv420p" is a pixel format using YUV color space and 4:2:0 chroma subsampling
+            # "-pix_fmt",
+            # "yuv420p",
```
Comment (anthonyjclark) [inline learning/inference.py:514]: Don't we need this one?

**Review by anthonyjclark — CHANGES_REQUESTED** [review]

File: boxnav/boxsim.py (hunk 10 above)
Comment (tonnyryam) [reply boxnav/boxsim.py:3]: I was using it to name the automatically name datasets when using the auto_upload functionality

File: boxnav/boxsim.py (hunk 11 above)
Comment (tonnyryam) [reply boxnav/boxsim.py:33]: I didn't think it was very helpful since boxsim would continue running based on the number of actions and not environment completion and it is used for data collection. It could be useful with smaller runs where you would be watching it the whole way through but not for when we were collecting the 100k image datasets for example or other large runs.

File: boxnav/boxsim.py (hunk 13 above)
Comment (tonnyryam) [reply boxnav/boxsim.py:72]: That is right, my mistake.

File: learning/inference.py (hunk 14 above)
Comment (tonnyryam) [reply learning/inference.py:10]: I don't think I pushed the changes yet but it is used for video creation in inference.

File: learning/inference.py (hunk 15 above)
Comment (tonnyryam) [reply learning/inference.py:514]: Inference video creation seems to work fine either way with or without it.

File: boxnav/boxsim.py (hunk 12 above)
Comment (tonnyryam) [reply boxnav/boxsim.py:82]: I used f-strings in other places as needed as well.
